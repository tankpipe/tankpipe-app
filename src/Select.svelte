<script>
    export let items = []
    export let item = {}
    export let label = null
    export let none = false
    export let flat = false
    export let inError = false
    export let disabled = false
    export let valueField = null
    export let onChange = {}
    const cssClass =  label ? "widget" : "widgetNoLabel"
</script>

<div class={cssClass}>
    {#if label}<label for="itemSelect">{label}</label>{/if}
    <select bind:value={item} name="itemSelect" class:flat={flat} class:error={inError} disabled={disabled} on:change={onChange}>
        {#if none}
            <option value={null}>None</option>
        {/if}
        {#each items as a}
            {#if valueField}
                <option value={a[valueField]}>{a.name}</option>
            {:else}
                <option value={a}>{a.name}</option>
            {/if}
        {/each}
    </select>
</div>

<style>
    .widget {
        display: inline-block;
        padding: 5px 0px 5px 10px;
    }

    .widgetNoLabel {
        display: inline-block;
    }

    label {
        text-align: left;
        font-size: .8em;
        margin-bottom: 3px;
    }

    .flat {
        background-color: #aaa;
        appearance: none;
        position: relative;
        background-image: url("data:image/svg+xml;utf8,<svg fill='black' height='24' viewBox='0 0 24 24' width='24' xmlns='http://www.w3.org/2000/svg'><path d='M7 10l5 5 5-5z'/><path d='M0 0h24v24H0z' fill='none'/></svg>");
        background-repeat: no-repeat;
        background-position-x: 100%;
        background-position-y: 5px;
        padding-right: 2rem;
    }

    .flat::after {
        content: "\25b6";
        color: #fff;
        position: absolute; top: 0; right: 0;
    }

    .error {
        border: 1px solid #FBC969 !important;
    }

    :global(.error-input input) {
        border: 1px solid #FBC969 !important;
    }
</style>