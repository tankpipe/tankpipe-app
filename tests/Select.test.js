import { render } from '@testing-library/svelte'
import Select from '../src/Select.svelte'

it('is displayed correctly', async () => {

    let items = []
    items.push({
        id: "a7bb5dd7-7787-404a-9678-00e6be32fe9e",
        name: "Account 1",
        account_type: "Debit",
	    balance: 0,
	    starting_balance: 0
    })

    let item = null
    const label = "Label 1"
    const select = render(Select, {item: item, items: items, label, none:true})

    expect(select.container.outerHTML).toMatchSnapshot();
});

it('shows in error', async () => {

    let items = []
    items.push({
        id: "a7bb5dd7-7787-404a-9678-00e6be32fe9e",
        name: "Account 1",
        account_type: "Debit",
	    balance: 0,
	    starting_balance: 0
    })

    let item = null
    const label = "Label 1"
    const select = render(Select, {item: item, items: items, label, none:true, inError:true})

    expect(select.container.outerHTML).toMatchSnapshot();
});

it('is displayed correctly with no label and no none', async () => {

    let items = []
    items.push({
        id: "a7bb5dd7-7787-404a-9678-00e6be32fe9e",
        name: "Account 1",
        account_type: "Debit",
	    balance: 0,
	    starting_balance: 0
    })

    let item = null
    const label = "Label 1"
    const select = render(Select, {item: item, items: items})

    expect(select.container.outerHTML).toMatchSnapshot();
});
