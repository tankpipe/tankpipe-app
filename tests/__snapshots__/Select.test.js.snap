// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`is displayed correctly 1`] = `"<div><div class=\\"widget svelte-vv6jek\\"><label for=\\"itemSelect\\" class=\\"svelte-vv6jek\\">Label 1</label> <select name=\\"itemSelect\\" class=\\"svelte-vv6jek\\"><option value=\\"\\">None</option><option value=\\"[object Object]\\">Account 1</option></select></div></div>"`;

exports[`is displayed correctly with no label and no none 1`] = `"<div><div class=\\"widgetNoLabel svelte-vv6jek\\"> <select name=\\"itemSelect\\" class=\\"svelte-vv6jek\\"><option value=\\"[object Object]\\">Account 1</option></select></div></div>"`;

exports[`shows in error 1`] = `"<div><div class=\\"widget svelte-vv6jek\\"><label for=\\"itemSelect\\" class=\\"svelte-vv6jek\\">Label 1</label> <select name=\\"itemSelect\\" class=\\"svelte-vv6jek error\\"><option value=\\"\\">None</option><option value=\\"[object Object]\\">Account 1</option></select></div></div>"`;
